Add a text-based circuit drawer

johannesjmeyer opened the pull request:
**Context:**
During QHACK, people requested a possibility to graphically inspect circuits.

**Description of the Change:**
This PR adds a text-based circuit drawer that renders the `CircuitGraph` in text format. It is accessible through the methods `CircuitGraph.draw` and `BaseQNode.draw`. The methods have an optional argument `show_variable_names=False` that indicates if the variable names should be drawn instead of the variable values.

An example is the following:
```
# Not parametrized
 0: ──RX(a)─────────╭C────RX(d)─────────╭C──╭C──╭C──────────╭C──╭SWAP───╭SWAP───┤ ⟨Y⟩
 1: ──RX(b)──Z──────╰X───╭RY(b)──RX(e)──├X──╰Z──│───╭X──╭C──│───│───────├SWAP───┤ Var[H]
 2: ──RY(c)──RX(f)───U0──│──────────────╰C──────╰Z──╰C──│───╰X──╰SWAP───│───────┤ Sample[X]
 3: ─────────────────────╰C──────RZ(b)──────────────────╰X───────RZ(b)──│──────╭┤ ⟨H0⟩
 4: ────────────────────────────────────────────────────────────────────╰C─────╰┤ ⟨H0⟩
U0 =
[[1. 0.]
 [0. 1.]]
H0 =
[[1. 0. 0. 0.]
 [0. 1. 0. 0.]
 [0. 0. 1. 0.]
 [0. 0. 0. 1.]]

# Parametrized
 0: ──RX(0.1)───────────╭C────RX(2.765)───────────╭C──╭C──╭C──────╭C──────╭SWAP─────╭SWAP───┤ ⟨Y⟩
 1: ──RX(0.2)──Z────────╰X───╭RY(0.2)────RX(0.5)──├X──╰Z──│───╭X──│───╭C──│─────────├SWAP───┤ Var[H]
 2: ──RY(0.3)──RX(0.6)───U0──│────────────────────╰C──────╰Z──╰C──╰X──│───╰SWAP─────│───────┤ Sample[X]
 3: ─────────────────────────╰C──────────RZ(0.2)──────────────────────╰X───RZ(0.2)──│──────╭┤ ⟨H0⟩
 4: ────────────────────────────────────────────────────────────────────────────────╰C─────╰┤ ⟨H0⟩
U0 =
[[1. 0.]
 [0. 1.]]
H0 =
[[1. 0. 0. 0.]
 [0. 1. 0. 0.]
 [0. 0. 1. 0.]
 [0. 0. 0. 1.]]
```

**Benefits:**
Circuits can now be drawn to enable visual debugging.

**Possible Drawbacks:**
None.

**Ideas for further Features**
* Right now, templates like `StronglyEntanglingLayers` render weirdly, because the circuit graph does not know about the block structure and thus pushes all gates as far back as possible. This could be alleviated by adding a command `qml.block()` or `qml.pragma('block')` that contains the information that in circuit drawing, this should be treated as a block.
This could actually be implemented quite easily by having an Operation `Pragma` acting on all wires (thus blocking the moving in the layering) that has an empty decomposition (thus being automatically replaced with "do nothing" on all devices). The only task is to remove it from the printed circuit.

antalszava commented on pennylane/circuit_drawer.py:
```diff
@@ -0,0 +1,916 @@
+# pylint: disable=too-many-branches,too-many-arguments,too-many-return-statements,too-many-statements,consider-using-enumerate,too-many-instance-attributes
+
+
+def _transpose(target_list):
```
This method seems to work in such a way, that some sublists might be truncated during transposing:
```python
In [1]: a = _transpose([[1,2], [3, 5, 7]])                                                                                                                                                                  

In [2]: a                                                                                                                                                                                                   
Out[2]: [[1, 3], [2, 5]]
```
This behaviour could be included here.

Also, giving the example of transposing a matrix might be useful (when reading for the first time, I was unsure if this would be the same thing, though it might just be me :sweat_smile: )

johannesjmeyer replied:
Actually, this whole `Grid` class is somewhat "bad" in the sense the I just want to have a 2D array. I don't remember why I did not use `ndarray(object)`. I guess there was a problem when you wanted to edit strings inside a numpy array.

I think I will refactor this more toroughly

antalszava commented on pennylane/circuit_drawer.py:
```diff
@@ -0,0 +1,916 @@
+            raw_grid = []
+
+        self.raw_grid = raw_grid
+        self.raw_grid_transpose = _transpose(raw_grid)
```
How about making this one a property?

```python
@property
def raw_grid_transpose(self):
    return _transpose(self.raw_grid)
```
This way it could also be avoided that each time the transposed ``raw_grid`` needs to be created.

johannesjmeyer replied:
It is intentionally not a property so that I only do a transpose operation (which is costly) when something inside the grid changes

antalszava commented on pennylane/circuit_drawer.py:
```diff
@@ -0,0 +1,916 @@
+        """Append the layers of another Grid to this Grid.
+
+        Args:
+            other_grid (pennylane.circuit_drawer.Grid): Grid whos layers will be appended
```
```suggestion
            other_grid (pennylane.circuit_drawer.Grid): Grid whose layers will be appended
```

antalszava commented on pennylane/circuit_drawer.py:
```diff
@@ -0,0 +1,916 @@
+        If the target element is not in the list it will be added to the list.
+
+        Args:
+            target_element (np.ndarray): The object whos index is to be returned
```
```suggestion
            target_element (np.ndarray): The object whose index is to be returned
```

josh146 commented on pennylane/qnodes/base.py:
```diff
@@ -336,13 +374,49 @@ def _construct(self, args, kwargs):
+            else:
+                kwarg_variable = Variable(0, name=variable_name, is_kwarg=True)
+
+            self.kwarg_vars[variable_name] = kwarg_variable
```
This is great, but not so sure it should be placed directly in construct (since construct is a very central method to the QNode). I would suggest either a private method that can be called here, or even an auxiliary function in this file.

josh146 replied:
It would also be good to have more commenting of the code above

johannesjmeyer replied:
I refactored this part.

josh146 commented on pennylane/qnodes/base.py:
```diff
@@ -360,15 +434,10 @@ def _construct(self, args, kwargs):
                 # it's ok to directly pass auxiliary arguments since the circuit is re-constructed each time
                 # (positional args must be replaced because parameter-shift differentiation requires Variables)
-                res = self.func(*arg_vars, **kwargs)
+                res = self.func(*self.arg_vars, **kwargs)
```
How come this has been changed to an attribute?

johannesjmeyer replied:
I don't remember. But I changed it back

antalszava replied:
Right now my understanding would be that this is because of the caching functionality

josh146 commented on pennylane/variable.py:
```diff
@@ -92,15 +92,27 @@ class Variable:
+            and self.idx == other.idx
+            and self.is_kwarg == other.is_kwarg
+            and self.mult == other.mult
+        )
```
Nice!

josh146 commented on pennylane/variable.py:
```diff
@@ -132,10 +144,35 @@ def val(self):
+        """Returns a string representation of the Variable.
+
+        Args:
+            show_name_only (bool, optional): Render the name instead of the value. Defaults to False.
```
No defaults needed

josh146 commented on tests/test_circuit_drawer.py:
```diff
@@ -0,0 +1,1519 @@
+        assert qfunc.draw(show_variable_names=True) == (
+            " 0: ──H──╭C─────────────────────────────╭C─────────╭┤ ⟨Z ⊗ Z⟩ \n"
+            + " 1: ─────╰RX(a)──Rot(w[0], w[1], w[2])──╰RX(-1*a)──╰┤ ⟨Z ⊗ Z⟩ \n"
+        )
```
These tests are super thorough!

antalszava commented on pennylane/circuit_drawer.py:
```diff
@@ -0,0 +1,916 @@
+        }.items():
+            ret = ret.replace(old, new)
+
+        return ret
```
I also agree with that one!

antalszava commented on pennylane/circuit_drawer.py:
```diff
@@ -0,0 +1,916 @@
+
+
+class CharSet(abc.ABC):
+    """Charset base class."""
```
Here as the ``CharSet`` and its subclasses are easily separable from the rest of the logic, it might even be considered to organize them into a separate file (as currently there are 4-5 classes structured in the same file). Considering creating a ``pennylane/circuit_drawer`` directory and having these files there could also be an option.

johannesjmeyer replied:
This is a good idea!

josh146 replied:
I agree! subpackages are always preferable to a single module when they get to about this size

antalszava commented on pennylane/circuit_drawer.py:
```diff
@@ -0,0 +1,916 @@
+    MIDDLE_MULTI_LINE_GATE_CONNECTOR = "├"
+    BOTTOM_MULTI_LINE_GATE_CONNECTOR = "╰"
+    EMPTY_MULTI_LINE_GATE_CONNECTOR = "│"
+    CONTROL = "C"
```
Just out of curiosity, did you consider this one for control qubits?
https://www.fileformat.info/info/unicode/char/25cf/browsertest.htm

johannesjmeyer replied:
Yes, it didn't look good

antalszava commented on pennylane/circuit_drawer.py:
```diff
@@ -0,0 +1,916 @@
+    resolution_dict = {
+        "PauliX": "X",
+        "CNOT": "X",
+        "Toffoli": "X",
```
This might be worth considering:
https://www.compart.com/en/unicode/U+2295

johannesjmeyer replied:
The problem here is that those special unicode characters usually do not exist in the monospace fonts, which makes the display of the circuit _really ugly_. At the end of the circuit it doesn't matter, inside it does:

```
 0: ──RX(a)───────────────────────╭C────RX(angles[0])───────────────────────────────╭C─────╭C─────╭C──╭C──────────╭C──╭SWAP───╭SWAP───┤ ⟨Y⟩
 1: ──RX(b)────────Z──────────────╰⊕───╭RY(b)──────────RX(4*angles[1])──╭RY(0.359)──├⊕──Z──│───Z──╰Z──│───╭⊕──╭C──│───│───────├SWAP───┤ Var[H]
 2: ──RY(1.889*c)──RX(angles[2])───U0──│────────────────────────────────│───────────╰C─────╰⊕─────────╰Z──╰C──│───╰⊕──╰SWAP───│───────┤ Sample[X]
 3: ───────────────────────────────────╰C──────────────RZ(b)────────────╰C────────────────────────────────────╰⊕───────RZ(b)──│──────╭┤ ⟨H0⟩
 4: ──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╰C─────╰┤ ⟨H0⟩
```

antalszava replied:
True that!

antalszava commented on pennylane/circuit_drawer.py:
```diff
@@ -0,0 +1,916 @@
+        if name in RepresentationResolver.resolution_dict:
+            name = RepresentationResolver.resolution_dict[name]
+
+        if op.name in self.control_wire_dict and wire in [
```
A comment could be added here alogn the lines of
```python
# Resolving the control qubits of a controlled operation
```

antalszava commented on pennylane/circuit_drawer.py:
```diff
@@ -0,0 +1,916 @@
+                self.charset.PIPE + n_str + self.charset.CROSSED_LINES + n_str + self.charset.PIPE
+            )
+
+        if op.name == "PolyXP":
```
It would be nice to create a separate method for this one (it's alone ~100 lines inside of the  ``operator_representation`` method).

johannesjmeyer replied:
Yes, I will refactor this part

antalszava commented on pennylane/circuit_drawer.py:
```diff
@@ -0,0 +1,916 @@
+        inserted_indices = []
+
+        for i in range(grid.num_layers):
+            layer_operators = _remove_duplicates(grid.layer(i))
```
At this point, I'm slightly lost, how come duplicates need to be removed here? What would be duplicated for the specific ``layer``?

antalszava replied:
How about breaking this one up somehow like as follows:
```python
def build_decoration_layer(self, decoration_layer, wires):
    sorted_wires = wires.copy()
    sorted_wires.sort()

    decoration_layer[sorted_wires[0]] = self.charset.TOP_MULTI_LINE_GATE_CONNECTOR

    for k in range(sorted_wires[0] + 1, sorted_wires[-1]):
        if k in sorted_wires:
            decoration_layer[k] = self.charset.MIDDLE_MULTI_LINE_GATE_CONNECTOR
        else:
            decoration_layer[k] = self.charset.EMPTY_MULTI_LINE_GATE_CONNECTOR

    decoration_layer[
        sorted_wires[-1]
    ] = self.charset.BOTTOM_MULTI_LINE_GATE_CONNECTOR
    return decoration_layer
```
And creating separate methods for ``separate=True`` and ``separate=False`` might also be a way to increase readability (though it might also just complicate things actually).

johannesjmeyer replied:
Thanks for that input! I took it up and also managed to get rid of `separate` entirely.

antalszava commented on pennylane/circuit_drawer.py:
```diff
@@ -0,0 +1,916 @@
+            grid (pennylane.circuit_drawer.Grid): Grid that holds the circuit information
+            representation_grid (pennylane.circuit_drawer.Grid): Grid that holds the string representations and into
+                which the decorations will be inserted
+            separate (bool): Insert decorations into separate layers
```
As per my current understanding, the following sentence would further elaborate on this (if so, could be added to the docstring, or something similar in content):
```
If ``True``, a separate layer containing the decorator is created for each unique layer operator.

johannesjmeyer replied:
I added a clarification.

johannesjmeyer commented on pennylane/circuit_drawer.py:
```diff
@@ -0,0 +1,916 @@
+        return ret
+
+
+class CharSet(abc.ABC):
```
I just don't want people to instantiate it, and I hoped that inheriting from `ABC` would realize that.

josh146 replied:
I think this will still be instantiable. `abc.ABC` only blocks abstract classes from being instantiated if decorated abstract methods exist that aren't being overwritten.

Consider the following:

```python
import abc

class Test(abc.ABC): 
    pass

class Test2(abc.ABC):

    @abc.abstractmethod
    def must_overwrite(self):
        pass
```

The first class will be instantiable, whereas the second class won't be:
```python
>>> t = Test()
>>> t = Test2()
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-6-0e60783dfdf4> in <module>
----> 1 t = Test2()

TypeError: Can't instantiate abstract class Test2 with abstract methods must_overwrite
```

Without abstract methods/properties, this is really more of a standard base class as opposed to an abstract base class (in Python terminology)

johannesjmeyer commented on pennylane/circuit_drawer.py:
```diff
@@ -0,0 +1,916 @@
+        return "_" + str(num)
+
+
+Charsets = {
```
I change it to `CHARSETS`

johannesjmeyer commented on pennylane/circuit_drawer.py:
```diff
@@ -0,0 +1,916 @@
+        "GaussianState": "Gaussian",
+        "QuadraticPhase": "QuadPhase",
+        "X": "x",
+        "P": "p",
```
Thanks for those short notations! Do you also have them for gates like `Kerr` and `CrossKerr`? I didn't manage to come up with my own.

johannesjmeyer commented on pennylane/circuit_drawer.py:
```diff
@@ -0,0 +1,916 @@
+
+        target_list.append(target_element)
+
+        return len(target_list) - 1
```
It could, but at that point in the code I know it's the last element because I just inserted it, so `len` is cheaper

johannesjmeyer commented on pennylane/circuit_drawer.py:
```diff
@@ -0,0 +1,916 @@
+
+        return " {} {} {}".format(
+            "+" if coefficient > 0 else "-", round(math.fabs(coefficient), 3), variable
+        )
```
I updated it and changed the styling a little!

johannesjmeyer commented on pennylane/circuit_drawer.py:
```diff
@@ -0,0 +1,916 @@
+        )
+
+    def operator_representation(self, op, wire):
+        """Resolve the representation of an Operator.
```
Yes, that's what I used it for in the entire PR. We could also use "Find" or "Get", what would you prefer?

josh146 replied:
Ah! I would prefer something very explicit, i.e., "Return the string representation of an Operator" (or some permutation thereof). Find and Get are also good :+1:

johannesjmeyer commented on pennylane/circuit_drawer.py:
```diff
@@ -0,0 +1,916 @@
+    Returns:
+        list[Hashable]: The input list without duplicate entries
+    """
+    return list(OrderedDict.fromkeys(input_list))
```
I guess this is even the most "canonical" way you can do it in python. Before I had `list(set(x))`, but that does not preserve order (I always wondered why my circuits render differently with each run...)

johannesjmeyer commented on pennylane/circuit_drawer.py:
```diff
@@ -0,0 +1,916 @@
+            representation_grid.append_layer(representation_layer)
+
+    def resolve_decorations(self, grid, representation_grid, separate):
+        """Resolve the decorations of the given Grid.
```
Everything that is not a gate or a wire. But up to now, this only includes the connectors for multi-wire gates.

johannesjmeyer commented on pennylane/circuit_graph.py:
```diff
@@ -288,18 +292,110 @@ def layers(self):
+        if list:
+            return list
+
+        return [None]
```
Actually, I got rid of the second one, it was unnecessary to have that as a method

johannesjmeyer commented on pennylane/qnodes/base.py:
```diff
@@ -234,6 +236,25 @@ def print_applied(self):
 
+    def draw(self, charset="unicode", show_variable_names=False):
+        """Draw the QNode as a circuit diagram.
+
```
Did it, I hope the way I added it renders correctly in the docs?

johannesjmeyer commented on pennylane/qnodes/base.py:
```diff
@@ -311,6 +332,23 @@ def _append_op(self, op):
                 )
             self.queue.append(op)  # TODO rename self.queue to self.op_queue
 
+    def _determine_variable_name(self, variable_value, prefix):
```
Added it!

smite commented on pennylane/qnodes/base.py:
```diff
@@ -234,6 +235,51 @@ def print_applied(self):
+        if self.circuit:
+            return self.circuit.draw(charset=charset, show_variable_names=show_variable_names)
+
+        raise QuantumFunctionError("The QNode can only be drawn after its CircuitGraph has been constructed.")
```
```suggestion
        raise RuntimeError("The QNode can only be drawn after its CircuitGraph has been constructed.")
```

smite replied:
`QuantumFunctionError` indicates that there's something wrong with the quantum function itself, so a `RuntimeError` might be better. Too bad there's no `UserError` :)

smite commented on pennylane/qnodes/base.py:
```diff
@@ -311,6 +357,113 @@ def _append_op(self, op):
+        or list structure.
+
+        Args:
+            parameter_value (Union[numeric,Sequence[Any],array[Any]]): The value of the parameter. This will be used as a blueprint for the returned variable name(s).
```
```suggestion
            parameter_value (Union[Number, Sequence[Any], array[Any]]): The value of the parameter. This will be used as a blueprint for the returned variable name(s).
```

smite commented on pennylane/qnodes/base.py:
```diff
@@ -311,6 +357,113 @@ def _append_op(self, op):
+        In this example, ``_make_variables`` will return the following :class:`~.variable.Variable` instances
+
+        .. code-block:: python
+            >>> dev._make_variables([3.4, [1.2, 3.4, 5.6]], {})
```
```suggestion
            >>> qfunc._make_variables([3.4, [1.2, 3.4, 5.6]], {})
```

smite commented on pennylane/qnodes/base.py:
```diff
@@ -336,13 +374,49 @@ def _construct(self, args, kwargs):
+        # pylint: disable=attribute-defined-outside-init, too-many-branches, too-many-statements
+
+        # flatten the args, replace each argument with a Variable instance carrying a unique index and name
+        full_argspec = inspect.getfullargspec(self.func)
```
Could this function call be replaced by adding more functionality to `_get_signature`, which is only called once when the BaseQNode is constructed?

smite replied:
The variable names never change (right?), so I think it'd make sense to build the name strings only once, not each time `_construct` is called. I think the same might apply for `arg_vars` and `kwarg_vars` in general, assuming the nested structure of the arguments isn't allowed to vary between qfunc calls?

johannesjmeyer replied:
I guess you're correct abouth the `getfullargspec`, that only has to be called once. But `arg_vars` and `kwarg_vars` can only be constructed during `_construct`, because without the user supplying arguments, we don't know their nested structure.

If one would enforce that parameters structure does not change (which is sensible in my opinion), we could cache the `arg_vars` and `kwarg_vars` and only construct them during the first pass.

josh146 replied:
@johannesjmeyer, that reminds me - does the circuit drawer construct code still work when `mutable=False`? QNodes that are immutable cache the result of the contract method and only ever call it once.

johannesjmeyer replied:
I see no reason why it should not. As long construct is called once everything is fine? The important thing is that the Variables end up in the circuit_graph and that happens during construction.

johannesjmeyer replied:
I implemented a caching logic. This also avoids the flattening step!

johannesjmeyer commented:
Thanks @smite for the suggestions!

josh146 commented on pennylane/circuit_drawer/charsets.py:
```diff
@@ -0,0 +1,148 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the different charsets supported by PennyLane's CircuitDrawer.
```
```suggestion
This module contains the different character set classes supported by PennyLane's CircuitDrawer.
```

josh146 commented on pennylane/circuit_drawer/charsets.py:
```diff
@@ -0,0 +1,148 @@
+    VERTICAL_LINE = None
+    CROSSED_LINES = None
+    PIPE = None
+    OTIMES = None
```
Small suggestion: in the abstract base class, you could also make the above abstract properties. This has two advantages:

1. inheriting classes are forced to overwrite them

2. the linter will then force you to add docstrings so that developers know where they are used :slightly_smiling_face: 

e.g., 

```python
@property
@abstractmethod
def WIRE(self):
    """The string to use to represent a wire"""
```

johannesjmeyer replied:
I implemented it that way!

josh146 commented on pennylane/circuit_drawer/charsets.py:
```diff
@@ -0,0 +1,148 @@
+    @staticmethod
+    def to_subscript(num):
+        """Convert the given number to a subscripted string."""
+        return "_" + str(num)
```
Something that just came to mind --- is the ASCII charset uneeded? In Python 3, all strings are unicode strings, so there should never be an environment where a user running PennyLane won't be able to output unicode?

johannesjmeyer replied:
I wouldn't bet on it. It could be that users run pennylane in a shell that can't display unicode characters or that they are running a monospace font that does not have those characters.

josh146 commented on pennylane/circuit_drawer/representation_resolver.py:
```diff
@@ -0,0 +1,368 @@
+
+        return "{:+.3g}{}".format(coefficient, variable)
+
+    def _format_polyxp(self, operation):
```
(docstring?)

josh146 commented on pennylane/circuit_drawer/representation_resolver.py:
```diff
@@ -0,0 +1,368 @@
+        if isinstance(element, qml.operation.Observable) and element.return_type is not None:
+            return self.output_representation(element, wire)
+
+        return self.operator_representation(element, wire)
```
Much easier to read and follow now that it is split into files!

josh146 commented on pennylane/circuit_graph.py:
```diff
@@ -346,18 +364,80 @@ def layers(self):
             yield LayerData(pre_queue, ops, tuple(param_inds), post_queue)
 
+    def greedy_layers(self):
+        """Greedily collected layers of the circuit. Empty slots are filled with ``None``.
```
It might be good to have an explanation here of how this differs from `parametrized_layers`

johannesjmeyer replied:
I actually don't exactly get what `parametrized_layers` does. Maybe you could add a more detailed docstring to that method? I'll expand the one for greedy layers.

josh146 commented on pennylane/circuit_graph.py:
```diff
@@ -346,18 +364,80 @@ def layers(self):
+
+        Returns:
+            Tuple[list[list[~.Operation]], list[list[~.Observable]]]:
+                Tuple of the circuits operations and the circuits observables, both indexed
```
```suggestion
            Tuple of the circuits operations and the circuits observables, both indexed
```

josh146 replied:
(since there is only ever one return type, continuation lines shouldn't be indented)

josh146 commented on pennylane/circuit_graph.py:
```diff
@@ -346,18 +364,80 @@ def layers(self):
+        Returns:
+            Tuple[list[list[~.Operation]], list[list[~.Observable]]]:
+                Tuple of the circuits operations and the circuits observables, both indexed
+                by wires.
```
```suggestion
            by wires.
```

josh146 commented on pennylane/qnodes/base.py:
```diff
@@ -234,6 +235,51 @@ def print_applied(self):
+        Consider the following circuit as an example:
+
+        .. code-block:: python
+            @qml.qnode(dev)
```
```suggestion

            @qml.qnode(dev)
```

josh146 commented on pennylane/qnodes/base.py:
```diff
@@ -234,6 +235,51 @@ def print_applied(self):
+        We can draw the circuit after it has been executed:
+
+        .. code-block:: python
+            >>> result = qfunc(2.3, [1.2, 3.2, 0.7])
```
```suggestion

            >>> result = qfunc(2.3, [1.2, 3.2, 0.7])
```

josh146 commented on pennylane/qnodes/base.py:
```diff
@@ -234,6 +235,51 @@ def print_applied(self):
+
+        Consider the following circuit as an example:
+
+        .. code-block:: python
```
```suggestion
        .. code-block:: python3
```

josh146 replied:
This distinction is needed here, as python (a shortcut to the python2 syntax highlighter) didn't have the `@` operator, so it will fail at the tensor observable

josh146 commented on pennylane/qnodes/base.py:
```diff
@@ -311,6 +357,113 @@ def _append_op(self, op):
+
+        The created :class:`~.variable.Variable` instances are given in the same nested structure
+        as the original arguments. The :class:`~.variable.Variable` instances are named according
+        to the python expression that adresses them inside the qfunc. Consider the following example:
```
```suggestion
        to the argument names given in the QNode definition. Consider the following example:
```

josh146 commented on pennylane/qnodes/base.py:
```diff
@@ -311,6 +357,113 @@ def _append_op(self, op):
+        as the original arguments. The :class:`~.variable.Variable` instances are named according
+        to the python expression that adresses them inside the qfunc. Consider the following example:
+
+        .. code-block:: python
```
```suggestion
        .. code-block:: python3
```

josh146 commented on pennylane/qnodes/base.py:
```diff
@@ -311,6 +357,113 @@ def _append_op(self, op):
+        to the python expression that adresses them inside the qfunc. Consider the following example:
+
+        .. code-block:: python
+            @qml.qnode(dev)
```
```suggestion

            @qml.qnode(dev)
```

josh146 commented on pennylane/qnodes/base.py:
```diff
@@ -311,6 +357,113 @@ def _append_op(self, op):
+
+                return qml.expval(qml.PauliZ(0) @ qml.PauliZ(1))
+
+            qfunc(3.4, [1.2, 3.4, 5.6])
```
Might look cleaner to move this below to the 'prompt' codeblock?

johannesjmeyer replied:
Yeah, you're right!

josh146 commented on pennylane/qnodes/base.py:
```diff
@@ -311,6 +357,113 @@ def _append_op(self, op):
+
+        In this example, ``_make_variables`` will return the following :class:`~.variable.Variable` instances
+
+        .. code-block:: python
```
```suggestion
        .. code-block:: python3

```

josh146 commented on pennylane/qnodes/base.py:
```diff
@@ -311,6 +357,113 @@ def _append_op(self, op):
+
+        .. code-block:: python
+            >>> qfunc._make_variables([3.4, [1.2, 3.4, 5.6]], {})
+            ["a", ["w[0]", "w[1]", "w[2]"]], {}
```
i.e.,

```rest
>>> qfunc(3.4, [1.2, 3.4, 5.6])
<some value>
>>> qfunc._make_variables([3.4, [1.2, 3.4, 5.6]], {})
["a", ["w[0]", "w[1]", "w[2]"]], {}

josh146 commented on pennylane/qnodes/base.py:
```diff
@@ -311,6 +357,113 @@ def _append_op(self, op):
+
+            kwarg_vars[variable_name] = kwarg_variable
+
+        return arg_vars, kwarg_vars
```
This makes it much easier to read and parse! Also, it's great that a developer-facing method has such a detailed docstring, it will make it much easier for other developers.

My only concern is the additional use of `_flatten()` above every time construct is called. `_flatten` can be _very_ slow for large arrays, so this could create a bottleneck/slowdown

johannesjmeyer replied:
I implemented Variable caching so that `make_variables` is only called once

josh146 commented on pennylane/qnodes/base.py:
```diff
@@ -336,13 +489,9 @@ def _construct(self, args, kwargs):
-        self.num_variables = len(arg_vars)
-        # arrange the newly created Variables in the nested structure of args
-        arg_vars = unflatten(arg_vars, args)
+        arg_vars, kwarg_vars = self._make_variables(args, kwargs)
```
Oh, I suppose the _flatten that used to be here has just moved into _make_variables :laughing:

josh146 commented on pennylane/variable.py:
```diff
@@ -132,10 +144,35 @@ def val(self):
+        if self.mult != 1:
+            return "{}*{}".format(str(round(self.mult, 3)), self.name)
+
+        return self.name
```
Should this be `__str__`, and the existing `__str__` becoming `__repr__`?

johannesjmeyer replied:
I have no hard feelings concerning this. But if the above thing would become `__str__` you would essentially get `w[0, 3]` instead of `Variable 0: name = w[0, 3]`. I guess it is not detrimental to show more info in `str` (as is the current approach)

johannesjmeyer commented:
Thanks for the re-review, @josh146, I implemented your suggestions. Furthermore, I refactored the `Grid` class to use a numpy array with dtype object as underlying raw grid. This should yield improved performance. I also added a caching logic for the `Variable`s in the base QNode.

antalszava commented on pennylane/circuit_drawer/__init__.py:
```diff
@@ -0,0 +1,26 @@
+    :toctree: api
+
+    default_qubit
+    default_gaussian
```
Would it be working with these two devices or was this part coming from another module docstring?

johannesjmeyer replied:
Oh sorry, that was a copy-paste error!

antalszava commented on pennylane/circuit_drawer/grid.py:
```diff
@@ -0,0 +1,193 @@
+        if self.raw_grid is not None:
+            return self.raw_grid.shape[1]
+
+        return 0
```
Could this become a one-liner?
```python
return 0 if self.raw_grid is None else self.raw_grid.shape[1]
```

johannesjmeyer replied:
It could, but I'd leave it as is. It makes no big difference and the current version is easier to read

antalszava commented on pennylane/circuit_drawer/grid.py:
```diff
@@ -0,0 +1,193 @@
+        if self.raw_grid is not None:
+            return self.raw_grid.shape[0]
+
+        return 0
```
Similar to the suggestion for ``num_layers``

antalszava commented on pennylane/circuit_drawer/representation_resolver.py:
```diff
@@ -0,0 +1,406 @@
+                    coefficients[y], "p{}".format(self.charset.to_subscript(idx))
+                )
+
+            return poly_str
```
Looks really nice that it has its own method! :blush: 

Another step could even be breaking it up into an ``order==1`` and ``order==2`` cases, if I see it correctly they would each require passing only the ``coefficients`` as an argument.

This way this bigger method would also be broken down.

johannesjmeyer replied:
That's a good suggestion, I implemented it!

antalszava commented on pennylane/circuit_drawer/representation_resolver.py:
```diff
@@ -0,0 +1,406 @@
+            return "cos({0})x+sin({0})p".format(par_rep)
+
+        if op.name == "FockStateProjector":
+            n_str = ",".join([str(n) for n in op.params[0]])
```
An idea here, since there are many ``if`` statements would be to create a ``dict``. This would have ``op.name`` as the key and the constructed ``string`` as the value. There are some values that are duplicated, but I guess it should be fine.

Then the return would look something like ``return control_representations[op.name]``.

johannesjmeyer replied:
I don't see how that would work. If one were to convert this to a dict it would have to be a dict with a lot of lambdas in it -- and that's surely less readable.
```python
string_representations = {
    "Hermitian" : lambda op: RepresentationResolver._format_matrix_operation(
                op, "H", self.hermitian_matrix_cache
            ),
    "etc" : ...
}
```

antalszava commented on pennylane/circuit_drawer/circuit_drawer.py:
```diff
@@ -0,0 +1,322 @@
+        self.operation_representation_grid = Grid()
+        self.observable_representation_grid = Grid()
+        self.operation_decoration_indices = []
+        self.observable_decoration_indices = []
```
(Very minor thing here: for me, it would be easier to read the initializations in the order of how the arguments were passed and then have a ``newline`` followed. This way it would be easy to follow how the attributes are created.)

johannesjmeyer replied:
:+1:

antalszava commented on pennylane/circuit_drawer/charsets.py:
```diff
@@ -0,0 +1,199 @@
+    EMPTY_MULTI_LINE_GATE_CONNECTOR = "│"
+    CONTROL = "C"
+    LANGLE = "⟨"
+    RANGLE = "⟩"
```
For me, these render not so well, and it is hard to tell the difference between them and the pair of ``(`` , ``)``. Could ``<`` and ``>`` work here as well?

johannesjmeyer replied:
For me they render fine. And I don't like those `<`, `>` because they don't really look like angled brackets because of their height

johannesjmeyer commented:
Thanks @antalszava for your re-review! I commented on your suggestions and implemented some of them. Thanks a lot for pointing out the problem with the unused wires. That was a little bit tricky to resolve and the solution does not involve displaying all wires. This is due to the fact that the `CircuitDrawer` is based on the `CircuitGraph` which only contains active wires.

antalszava commented on tests/circuit_drawer/test_circuit_drawer.py:
```diff
@@ -0,0 +1,647 @@
+        """Test that a qubit circuit with unused wires renders correctly."""
+        output = qubit_circuit_with_unused_wires.draw()
+
+        assert output == drawn_qubit_circuit_with_unused_wires
```
Really nice, thanks for adding this one!

antalszava commented on pennylane/circuit_drawer/circuit_drawer.py:
```diff
@@ -116,6 +121,14 @@ def __init__(
 
+    def make_wire_conversion_dicts(self, raw_operation_grid, raw_observable_grid):
+        all_operators = list(qml.utils._flatten(raw_operation_grid)) + list(qml.utils._flatten(raw_observable_grid))
+        all_wires = [op.wires for op in all_operators if op is not None]
```
Just out of curiosity, when does the ``op is None`` case arise?

johannesjmeyer replied:
In the grid, places where no operation is taking place are filled with `None`.

antalszava commented on tests/circuit_drawer/test_representation_resolver.py:
```diff
@@ -0,0 +1,711 @@
+
+        unicode_representation_resolver.element_representation(op, wire)
+
+        print()
```
```suggestion

```

antalszava commented on tests/circuit_drawer/test_representation_resolver.py:
```diff
@@ -0,0 +1,711 @@
+        wire = 3
+
+        unicode_representation_resolver.element_representation(op, wire)
+
```
```suggestion
```

antalszava commented on pennylane/circuit_drawer/circuit_drawer.py:
```diff
@@ -0,0 +1,372 @@
+        internal_wires = list(range(len(device_wires)))
+
+        self._device_wire_to_internal_wire = dict(zip(device_wires, internal_wires))
+        self._internal_wire_to_device_wire = dict(zip(internal_wires, device_wires))
```
Could the ``inv_map = {v: k for k, v in my_map.items()}`` type of inverting a dictionary work here, once you have ``self._device_wire_to_internal_wire``? This mapping seems to be one-to-one, so it seems that there shouldn't be a problem with inversion.

johannesjmeyer replied:
Of course one could do it like this, but as I have the keys and values at hand I prefer to do it this way around because it just reads easier.

antalszava commented on pennylane/circuit_drawer/circuit_drawer.py:
```diff
@@ -0,0 +1,372 @@
+        internal_wires = list(range(len(device_wires)))
+
+        self._device_wire_to_internal_wire = dict(zip(device_wires, internal_wires))
+        self._internal_wire_to_device_wire = dict(zip(internal_wires, device_wires))
```
One thing to note here is that ``device`` seems to indicate that you get the ``wires`` from the ``device``. Whereas they are received from the ``QNode`` (if I'm not mistaken). Could using ``circui_wires`` be better here?

johannesjmeyer replied:
Let's use circuit wires!
